File: Question/Question 31-40/32. Longest Valid Parentheses.java
import java.util.Stack;

class aa {
    public int longestValidParentheses(String s) {
        Stack<Integer> stack = new Stack<Integer>();
        stack.add(-1);
        int maxNum = 0;
        for(int i=0; i<s.length(); i++){
            if(s.charAt(i) == '('){
                stack.add(i);
            }else {
                stack.pop();
                if(stack.isEmpty()){
                    stack.add(i);
                }
                maxNum = Math.max(maxNum, i - stack.peek());
            }
        }
        return maxNum;
    }
}

File: Question/Question 31-40/32. Longest Valid Parentheses.js
/**
 * @param {string} s
 * @return {number}
 */
let longestValidParentheses = function(str){
	let arr = str.split("");
    let stack =[-1];
    let count=0;
    for(let i=0;i<arr.length;i++){
        if(arr[i]==="("){
            stack.push(i);
        }else{
            stack.pop();
            if(stack.length === 0){
                stack.push(i);
            }
            count = Math.max(count, i-stack[stack.length-1]);
        }
    }
    return count;
}
console.log(longestValidParentheses("(()"));
console.log(longestValidParentheses(")()())"));
console.log(longestValidParentheses("()(()"));
console.log(longestValidParentheses("()(())"));